### PR TITLE
project-mongo-api-Therese Linden

### PR DESCRIPTION
**Pull request by thereselinden**
**Link:** 
https://thereselinden-mongo-book-api.herokuapp.com/

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,35 @@
 # Mongo API Project
```
Comment (ylvan75) [inline README.md:1]: Clean and nice README.
- Easy to read

File: server.js [hunk 2]
```diff
@@ -1,38 +1,195 @@
...
+import cors from 'cors';
+import mongoose from 'mongoose';
+import booksData from './data/books.json';
+import authorsData from './data/authors.json';
+
+const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/project-mongo";
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true });
+mongoose.Promise = Promise;
+
+//_____________create modules
```
Comment (ylvan75) [inline server.js:12]: I really like these comments, so clean.

File: server.js [hunk 3]
```diff
@@ -1,38 +1,195 @@
...
-app.use(cors())
-app.use(bodyParser.json())
-
-// Start defining your routes here
+import express from 'express';
+import bodyParser from 'body-parser';
+import cors from 'cors';
+import mongoose from 'mongoose';
+import booksData from './data/books.json';
+import authorsData from './data/authors.json';
```
Comment (ylvan75) [inline server.js:6]: I think we should be proud of that we did our own data file with uniques authors.

File: server.js [hunk 4]
```diff
@@ -1,38 +1,195 @@
...
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true });
+mongoose.Promise = Promise;
+
+//_____________create modules
+const Book = mongoose.model('Book', {
+  bookID: Number,
+  title: String,
+  authors: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: 'Author',
```
Comment (ylvan75) [inline server.js:18]: Here as well, we should be proud of combine two collections

File: server.js [hunk 5]
```diff
@@ -1,38 +1,195 @@
...
+    .send({ error: "No books found"});
+  };
+});
+
+//_____________Single book
+app.get('/books/:id', async (req, res) => {
+  const { id } = req.params;
+
+  try {
+  const singleBook = await Book.findOne({ bookID: +id });
```
Comment (ylvan75) [inline server.js:121]: Seems good to make sure the id is a number by doing `+id`

File: server.js [hunk 6]
```diff
@@ -1,38 +1,195 @@
...
+  const pageNumber = +page || 1; 
+  const pageSize = 20;
+  const skip = pageSize * (pageNumber -1);
+
+  const authorsArray = await Author.find({
+    authors: new RegExp(author, 'i'),
+  }) 
+  .sort({ authors: 1 })
+  .limit(pageSize)
+  .skip(skip);;
```
Comment (ylvan75) [inline server.js:146]: Nice that you re-use the same structure as for the endpoint /books
- Sorting and pagination

File: server.js [hunk 7]
```diff
@@ -1,38 +1,195 @@
...
+  res.send({ 
+    currentPage: pageNumber,
+    numberOfResults: authorsArray.length, 
+    results: authorsArray 
+  });
+}); 
+
+//_____________Single author based in ID
+app.get('/authors/:id', async (req, res) => {
+  const singleAuthor = await Author.findById(req.params.id);
```
Comment (ylvan75) [inline server.js:162]: Just a minor thing to be consistent. Much easier to see these things on others code then being good at it myself.

```suggestion
const { id } = req.params;
const singleAuthor = await Author.findById({ _id: id });
```

File: server.js [hunk 8]
```diff
@@ -1,38 +1,195 @@
...
+  } else {
+    res
+    .status(404)
+    .json({ error: 'Author not found' });
+  }
+});
+
+//_____________Books created by author based on authors ID
+app.get('/authors/:id/books', async (req, res) => {
+  const author = await Author.findById(req.params.id);
```
Comment (ylvan75) [inline server.js:174]:
```suggestion
  const { id } = req.params;
const singleAuthor = await Author.findById({ _id: id });
```

File: server.js [hunk 9]
```diff
@@ -1,38 +1,195 @@
...
+  if (author) {
+    const books = await Book.find({
+      authors: mongoose.Types.ObjectId(author.id)
+    })
+    .populate('authors')
+    .sort({ average_rating: -1 });
```
Comment (ylvan75) [inline server.js:180]: I like this!

File: server.js [hunk 10]
```diff
@@ -1,38 +1,195 @@
...
+    res.json({ 
+      numberOfBooks: books.length, 
+      results: books 
+    });
+  } else {
+    res
+    .status(404)
+    .json({ error: 'authors/:id/books' });
```
Comment (ylvan75) [inline server.js:188]: Seems like you forgot to change the error message.

Perhaps you want something like this.

```suggestion
    .json({ error: `No books found with author id ${id}.` });
```

**Review by ylvan75 — APPROVED** [review]
Hi Therese, let's see if I can give you some good coding feedback. Since you and I worked together during this project and therefore of course we have chosen the same code solution for our endpoints I think it will be hard.

I start by saying that I like both your README and server.js, they are easy to read. I really like your way of doing comments.

**Things that I think we should be proud off that we managed to do/figure out:**

- Combine two collections

- Created our own data file with uniques authors - authors.json
--> Here we used books.json and with code (unfortunately I don't remember exactly how we did) we were able to filter out unique authors. That each author was an object in an array.

- How to get Heroku that runs our server.js to populate our datebase on MongoDB Cloud Atlas

It was really tough to do the code review and came up with good code improvements. I hope one of the code coaches will give us more in depth code improvements later on to reflect on and improve our backend skills.

**Two things for us to keep in mind for future projects:**

- Use const's instead of text in error handling

- To better understand when to use json or send, I saw that we both did json for the error handling for our last endpoint --> /authors/:id/books'

As always fun working with you Therese!

File: server.js [hunk 11]
```diff
@@ -1,38 +1,195 @@
...
+mongoose.Promise = Promise;
+
+//_____________create modules
+const Book = mongoose.model('Book', {
+  bookID: Number,
+  title: String,
+  authors: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: 'Author',
+  },
```
Comment (KarolCislowski) [inline server.js:19]: 🥇 relation to another model 😻

File: server.js [hunk 12]
```diff
@@ -1,38 +1,195 @@
...
+    .send({ error: `${id} is not a valid bookID`});
+  };
+});
+
+//_____________Array of authors
+app.get("/authors", async (req, res) => {
+  const { author, page } = req.query;
+  const pageNumber = +page || 1; 
+  const pageSize = 20;
+  const skip = pageSize * (pageNumber -1);
```
Comment (KarolCislowski) [inline server.js:139]: You used this logic again by copying the code. You should create a separate function to perform this logic and call it when needed.

**Review by KarolCislowski — APPROVED** [review]
Great Project! 🥇

**Pull request closed without merging**